File: app/src/main/java/com/kerimovscreations/lateandroid/activities/MainActivity.java
package com.kerimovscreations.lateandroid.activities;

import android.annotation.SuppressLint;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.TimePickerDialog;
import android.os.Build;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.view.LayoutInflater;
import android.view.View;
import android.view.animation.AnimationSet;
import android.view.animation.LinearInterpolator;
import android.view.animation.RotateAnimation;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.res.ResourcesCompat;
import androidx.work.Data;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

import com.kerimovscreations.lateandroid.R;
import com.kerimovscreations.lateandroid.dialogs.ReminderPickerDialogFragment;
import com.kerimovscreations.lateandroid.models.ReminderOption;
import com.kerimovscreations.lateandroid.tools.HelpFunctions;
import com.kerimovscreations.lateandroid.workers.NotifyWorker;

import java.util.ArrayList;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainActivity extends AppCompatActivity {

    public static final String workTag = "notificationWork";

    // Views

    @BindView(R.id.timer_circle)
    ImageView mTimerCircleImage;
    @BindView(R.id.btn_play)
    ImageView mPlayBtn;
    @BindView(R.id.timer_time_min)
    TextView mTimerTimeText;
    @BindView(R.id.timer_time_sec)
    TextView mTimerSecondText;

    // Variables

    private CountDownTimer timer;
    private long mTimerDuration;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        initVars();
    }

    void initVars() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CharSequence name = "Reminders";
            String description = "Reminders for time periods of meetings";
            int importance = NotificationManager.IMPORTANCE_HIGH;
            NotificationChannel channel = new NotificationChannel("B", name, importance);
            channel.setDescription(description);

            channel.enableLights(true);
            channel.enableVibration(true);
            channel.setSound(null, null);

            NotificationManager notificationManager = getSystemService(NotificationManager.class);

            if (notificationManager != null) {
                notificationManager.createNotificationChannel(channel);
            }
        }

        checkRecentTimer();
    }

    void checkRecentTimer() {
        if (HelpFunctions.shared.isRecentTimerActive(this)) {
            mTimerDuration = HelpFunctions.shared.getTimerDuration(this);
            long timerTimestamp = HelpFunctions.shared.getTimerStartTimestamp(this);
            long currentTimestamp = HelpFunctions.shared.getCurrentTimestamp();
            long millisUntilFinished = timerTimestamp + mTimerDuration - currentTimestamp;
            startTimer(millisUntilFinished);
            startTimerCircleAnimation(millisUntilFinished);
        }
    }

    /* Click handlers */

    @OnClick(R.id.btn_play)
    void onPlay() {
        if (this.timer != null) {
            stopTimer();
            deleteReminders();
        } else {
            pickTime();
        }
    }

    /* Functions */

    void pickTime() {
        TimePickerDialog timePickerDialog = new TimePickerDialog(this,
                (timePicker, hours, minutes) -> pickReminders(hours * 60 + minutes),
                0, 0, true);

        LayoutInflater inflater = getLayoutInflater();
        @SuppressLint("InflateParams") View view = inflater.inflate(R.layout.dialog_title, null);
        timePickerDialog.setCustomTitle(view);
        timePickerDialog.show();
    }

    void pickReminders(int minutes) {
        if (minutes > 0) {
            ReminderPickerDialogFragment dialogFragment = new ReminderPickerDialogFragment(this, minutes);
            dialogFragment.setOnInteractionListener(options -> {
                mTimerDuration = minutes * 60000;

                setReminders(mTimerDuration, options);

                HelpFunctions.shared.setTimerStartTimestamp(this, HelpFunctions.shared.getCurrentTimestamp());
                HelpFunctions.shared.setTimerDuration(this, mTimerDuration);

                startTimer(mTimerDuration);
                startTimerCircleAnimation(mTimerDuration);
            });
            dialogFragment.show();
        }
    }

    void startTimer(long duration) {
        mPlayBtn.setImageDrawable(ResourcesCompat.getDrawable(getResources(), R.drawable.ic_stop_white_24dp, null));

        timer = new CountDownTimer(duration, 1000) {
            public void onTick(long millisUntilFinished) {
                mTimerTimeText.setText(String.format(Locale.getDefault(), "%d %02d",
                        millisUntilFinished / 3600000 % 24,
                        millisUntilFinished / 60000 % 60));
                mTimerSecondText.setText(String.format(Locale.getDefault(), "%02d",
                        millisUntilFinished / 1000 % 60));
            }

            public void onFinish() {
                stopTimer();
            }
        };

        timer.start();
    }

    void startTimerCircleAnimation(long millisUntilFinished) {
        AnimationSet animSet = new AnimationSet(true);
        animSet.setInterpolator(new LinearInterpolator());
        animSet.setFillAfter(true);
        animSet.setFillEnabled(true);

        final RotateAnimation animRotate = new RotateAnimation(
                (float) ((double) (mTimerDuration - millisUntilFinished) / (double) mTimerDuration) * 178.0f,
                178.0f,
                RotateAnimation.RELATIVE_TO_SELF, 0.5f,
                RotateAnimation.RELATIVE_TO_SELF, 0.5f);

        animRotate.setDuration(mTimerDuration);
        animRotate.setFillAfter(true);
        animSet.addAnimation(animRotate);

        mTimerCircleImage.startAnimation(animSet);
    }

    @SuppressLint("SetTextI18n")
    void stopTimer() {
        this.timer.cancel();
        this.timer = null;
        this.mPlayBtn.setImageDrawable(ResourcesCompat.getDrawable(getResources(), R.drawable.ic_play_arrow_white_24dp, null));
        mTimerTimeText.setText("0 00");
        mTimerSecondText.setText("00");
        mTimerCircleImage.clearAnimation();
        HelpFunctions.shared.setTimerStartTimestamp(this, 0);
    }

    void setReminders(long duration, ArrayList<ReminderOption> options) {
        for (ReminderOption option : options) {
            if (!option.isSelected()) {
                continue;
            }

            setNotification(duration, option.getValue());
        }
    }

    void deleteReminders() {
        WorkManager.getInstance().cancelAllWorkByTag(workTag);
    }

    void setNotification(long duration, int value) {
        Data inputData = HelpFunctions.shared.getNotificationData(this, value);

        OneTimeWorkRequest notificationWork = new OneTimeWorkRequest.Builder(NotifyWorker.class)
                .setInitialDelay(duration - (value * 60000), TimeUnit.MILLISECONDS)
                .setInputData(inputData)
                .addTag(workTag)
                .build();

        WorkManager.getInstance().enqueue(notificationWork);
    }
}

File: app/src/main/java/com/kerimovscreations/lateandroid/activities/MainActivity.kt
package com.kerimovscreations.lateandroid.activities

import android.Manifest
import android.annotation.SuppressLint
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.CountDownTimer
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.res.ResourcesCompat
import androidx.databinding.DataBindingUtil
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkManager
import com.kerimovscreations.lateandroid.R
import com.kerimovscreations.lateandroid.models.ReminderOption
import com.kerimovscreations.lateandroid.databinding.ActivityMainBinding
import com.kerimovscreations.lateandroid.dialogs.GuidelinesDialogFragment
import com.kerimovscreations.lateandroid.dialogs.ReminderPickerDialogFragment
import com.kerimovscreations.lateandroid.dialogs.SettingsDialogFragment
import com.kerimovscreations.lateandroid.tools.BaseActivity
import com.kerimovscreations.lateandroid.tools.CircularSeekBar
import com.kerimovscreations.lateandroid.tools.HelpFunctions
import com.kerimovscreations.lateandroid.workers.NotifyWorker
import java.util.Locale
import java.util.concurrent.TimeUnit

class MainActivity : BaseActivity() {

    private lateinit var binding: ActivityMainBinding

    private var timer: CountDownTimer? = null
    private var mTimerDuration: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        initVars()
    }

    private fun initVars() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name: CharSequence = "Reminders"
            val description = "Audio reminders based on user settings"
            val importance = NotificationManager.IMPORTANCE_HIGH
            val channel = NotificationChannel("B", name, importance)
            channel.description = description
            channel.enableLights(true)
            channel.enableVibration(true)
            channel.setSound(null, null)
            val notificationManager = getSystemService(NotificationManager::class.java)
            notificationManager?.createNotificationChannel(channel)
        }

        checkRecentTimer()

        binding.btnPlay.setOnClickListener {
            if (timer != null) {
                stopTimer()
                deleteReminders()
            } else {
                if (hasNotificationPermission(this).not()) {
                    Toast.makeText(
                        this,
                        getString(R.string.warning_no_permission),
                        Toast.LENGTH_SHORT
                    ).show()
                    requestPushNotificationPermission()
                    return@setOnClickListener
                }

                val minutes = binding.circularSeekBar.progress
                if (minutes > 5) {
                    pickReminders(minutes)
                } else {
                    Toast.makeText(this, getString(R.string.warning_min_time), Toast.LENGTH_SHORT)
                        .show()
                }
            }
        }

        binding.btnSettings.setOnClickListener {
            val fragment = SettingsDialogFragment.newInstance()
            fragment.show(supportFragmentManager, "settings")
        }

        binding.btnInfo.setOnClickListener {
            val fragment = GuidelinesDialogFragment.newInstance()
            fragment.show(supportFragmentManager, "guidelines")
        }

        binding.circularSeekBar.setOnSeekBarChangeListener(object :
            CircularSeekBar.OnCircularSeekBarChangeListener {
            override fun onProgressChanged(
                circularSeekBar: CircularSeekBar?,
                progress: Int,
                fromUser: Boolean
            ) {
                updateTimeText((progress * 60).toLong())
            }

            override fun onStopTrackingTouch(seekBar: CircularSeekBar?) {
            }

            override fun onStartTrackingTouch(seekBar: CircularSeekBar?) {
            }
        })
    }

    /**
     * Functions
     */

    private fun checkRecentTimer() {
        if (HelpFunctions.shared.isRecentTimerActive(this)) {
            mTimerDuration = HelpFunctions.shared.getTimerDuration(this)
            val timerTimestamp = HelpFunctions.shared.getTimerStartTimestamp(this)
            val currentTimestamp = HelpFunctions.shared.currentTimestamp
            val millisUntilFinished = timerTimestamp + mTimerDuration - currentTimestamp
            startTimer(millisUntilFinished)
        }
    }

    private fun hasNotificationPermission(context: Context): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.POST_NOTIFICATIONS
            ) == PackageManager.PERMISSION_GRANTED
        } else {
            true
        }
    }

    private fun requestPushNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.POST_NOTIFICATIONS),
                1
            )
        }
    }

    private fun pickReminders(minutes: Int) {
        val dialogFragment = ReminderPickerDialogFragment.newInstance(minutes)
        dialogFragment.setOnInteractionListener(object :
            ReminderPickerDialogFragment.OnInteractionListener {
            override fun onSubmit(options: ArrayList<ReminderOption>) {
                mTimerDuration = minutes * 60000.toLong()
                setReminders(mTimerDuration, options)
                HelpFunctions.shared.setTimerStartTimestamp(
                    this@MainActivity,
                    HelpFunctions.shared.currentTimestamp
                )
                HelpFunctions.shared.setTimerDuration(this@MainActivity, mTimerDuration)
                startTimer(mTimerDuration)
            }
        })

        dialogFragment.show(supportFragmentManager, "")
    }

    private fun startTimer(duration: Long) {
        binding.btnPlay.setImageDrawable(
            ResourcesCompat.getDrawable(
                resources,
                R.drawable.ic_stop_white_24dp,
                null
            )
        )
        binding.circularSeekBar.isTouchEnabled = false
        timer = object : CountDownTimer(duration, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                updateTimeText(millisUntilFinished / 1000, true)
            }

            override fun onFinish() {
                stopTimer()
            }
        }
        timer?.start()
    }

    private fun updateTimeText(seconds: Long, shouldUpdateSeekBar: Boolean = false) {
        binding.timerTimeMin.text = String.format(
            Locale.getDefault(), "%d %02d",
            seconds / 3600 % 24,
            seconds / 60 % 60
        )
        binding.timerTimeSec.text = String.format(
            Locale.getDefault(), "%02d",
            seconds % 60
        )

        if (shouldUpdateSeekBar) {
            binding.circularSeekBar.progress = (seconds / 60).toInt()
        }
    }

    @SuppressLint("SetTextI18n")
    fun stopTimer() {
        timer?.cancel()
        timer = null
        binding.btnPlay.setImageDrawable(
            ResourcesCompat.getDrawable(
                resources,
                R.drawable.ic_play_arrow_white_24dp,
                null
            )
        )
        binding.timerTimeMin.text = "0 00"
        binding.timerTimeSec.text = "00"
        binding.circularSeekBar.progress = 0
        HelpFunctions.shared.setTimerStartTimestamp(this, 0)
        binding.circularSeekBar.isTouchEnabled = true
    }

    private fun setReminders(duration: Long, options: ArrayList<ReminderOption>) {
        for (option in options) {
            if (!option.isSelected) {
                continue
            }
            setNotification(duration, option.value)
        }
    }

    private fun deleteReminders() {
        WorkManager.getInstance(this).cancelAllWorkByTag(TAG_WORK)
    }

    private fun setNotification(duration: Long, value: Int) {
        val inputData = HelpFunctions.shared.getNotificationData(this, value)
        val notificationWork = OneTimeWorkRequest.Builder(NotifyWorker::class.java)
            .setInitialDelay(duration - value * 60000, TimeUnit.MILLISECONDS)
            .setInputData(inputData)
            .addTag(TAG_WORK)
            .build()
        WorkManager.getInstance(this).enqueue(notificationWork)
    }

    companion object {
        const val TAG_WORK = "notificationWork"
    }
}